### PR TITLE
Better versioning for BOUT++

### PR DESCRIPTION
**Pull request by dschwoerer**
Inspired by `setuptools_scm`

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,9 +16,24 @@ endif()
 
 # CMake currently doesn't support proper semver
 # Set the version here, strip any extra tags to use in `project`
-set(BOUT_FULL_VERSION 5.0.0-alpha)
-string(REGEX REPLACE "^([0-9]+\.[0-9]+\.[0-9]+)-.*" "\\1" BOUT_CMAKE_ACCEPTABLE_VERSION ${BOUT_FULL_VERSION})
-string(REGEX REPLACE "^[0-9]+\.[0-9]+\.[0-9]+-(.*)" "\\1" BOUT_VERSION_TAG ${BOUT_FULL_VERSION})
+# We try to use git to git a full description, inspired by setuptools_scm
```
Comment (d7919) [inline CMakeLists.txt:19]:
```suggestion
# We try to use git to get a full description, inspired by setuptools_scm
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -16,9 +16,24 @@ endif()
...
+# Not we use distance from 4.0.0 so it does not reset on minor releases.
```
Comment (d7919) [inline CMakeLists.txt:20]:
```suggestion
# Note we use distance from 4.0.0 so it does not reset on minor releases.
```

**Comment by dschwoerer** [comment]
Thanks @d7919

**Review by bendudson — DISMISSED** [review]

**Comment by dschwoerer** [comment]
I actually tried `setuptools_scm` - but that would only bump to 4.4.$(n+1).dev - and there is no option to get 5.0.0.dev - only 5.1.0 or 5.0.1.dev, haven't tested, not sure. The issue is here: https://github.com/pypa/setuptools_scm/issues/523
But that wouldn't work nice, because if we than merge 4.5 into next, it would use that instead of 5.0.0dev, Anyway, I wouldn't know how to programmatically decide on whether this is going to 5.0.0 or 4.5.0 or 4.4.x ...

So, yes, unfortunately, it needs manual bumping.

I think setting BOUT_PREVIOUS_VERSION and BOUT_NEXT_VERSION might be nicer :+1:

**Comment by ZedThree** [comment]
Ah, good point. That is a bit of a shame. Also very true that we just decide what the next version will be, and don't really use an algorithm. Nevermind, this is still nice!

**Comment by ZedThree** [comment]
I've changed the hardcoded versions to variables, and updated the release checklist to reflect this.

I did consider making the variables settable from the command line (by making them cache variables), which would make it easier to test and change the variables for maintainers, but would probably cause more problems than it would solve when the version doesn't update automatically.

**Comment by ZedThree** [comment]
This can go in as soon as the tests complete

**Pull request merged**